network/retrieve/retrieve.go

```diff
@@ -368,13 +368,17 @@ func (r *Retrieval) RequestFromPeers(ctx context.Context, req *storage.Request,
 	log.Debug("retrieval.requestFromPeers", "req.Addr", req.Addr)
 	metrics.GetOrRegisterCounter("network.retrieve.requestfrompeers", nil).Inc(1)
 
+FINDPEER:
 	sp, err := r.findPeer(ctx, req)
 	if err != nil {
 		log.Trace(err.Error())
 		return nil, err
 	}
 
 	protoPeer := r.getPeer(sp.ID())
+	if protoPeer == nil {
+		goto FINDPEER
```

Comment:
Potential endless loop. And looks like a very likely one, to me.

Reply:
If `getPeer` returns nil, it means that the connection is dropped and the peer is removed from the collection, so just return an error and that's it.

Reply:
guys, `RequestFromPeer` is triggered when we want a chunk from our peers. that is - we want a chunk, or another node asks us for a chunk and we can't deliver so we ask other peers.
the possible race condition that a peer that is returned from the kademlia in `r.findPeer` then disconnecting before we reach `r.getPeer` is so remote that i would barely even consider it feasible, also, the potential `very likely` infinite loop you're talking about means that every subsequent call to `findPeer` will yield the same race condition.that is very _un_ -likely.
returning an error when this race condition has occured makes no sense to me. why should we fail a delivery forwarding request from a certain peer just because another peer disconnected from us? we should just find another peer to request from.
my 2 cents

Reply:
@acud you're right, we should just find another suggested peer, as you've written it, I was confused when I wrote my comment, it makes sense now.

Reply:
My reasoning is with this code:
- findPeer returns no error
- getPeer returns  nil
-> infinite loop

Why findPeer and getPeer would behave like that, It does not matter to me in the context of this function, as with any change to this two functions it is possible, however unlikely it is to happen now. When someone changes this two functions, he/she should be aware of RequestFromPeers implications. If it is possible, my opinion is that it should be handled.

I just wanted to express my opinion. I am deeply sorry if I am frustrating anybody with this. Whatever you think is correct is ok with me.

Reply:
janos i have addressed this with a limit to how many retries we should do